In [1]:
import pandas as pd
import plotly_express as px 
import plotly.graph_objects as go

Athlete_events = pd.read_csv("Data/anonymized_olympics_data.csv")
Noc_regions = pd.read_csv("Data/noc_regions.csv")

df_athlete = pd.DataFrame(Athlete_events)
df_regions = pd.DataFrame(Noc_regions)
df_athlete.head()

Unnamed: 0  ID                                               Name Sex  \
0           0   1  3a4eef48434c66b3f14ab0221f6762d0ef7c6135ab2790...   M   
1           1   2  a6430cc6630934275dc6283f7e97e9625e6587cdddec7a...   M   
2           2   3  9c198b205332c2c8e1542e0f9534b9e270780a41d978ec...   M   
3           3   4  0a477bb1c5ad39716f9c775e54d18d16aa8b37ada55548...   M   
4           4   5  5b7be356aa28178096dc6747f0b8e4e393eaceb5f95310...   F   

    Age  Height  Weight            Team  NOC        Games  Year  Season  \
0  24.0   180.0    80.0           China  CHN  1992 Summer  1992  Summer   
1  23.0   170.0    60.0           China  CHN  2012 Summer  2012  Summer   
2  24.0     NaN     NaN         Denmark  DEN  1920 Summer  1920  Summer   
3  34.0     NaN     NaN  Denmark/Sweden  DEN  1900 Summer  1900  Summer   
4  21.0   185.0    82.0     Netherlands  NED  1988 Winter  1988  Winter   

        City          Sport                             Event Medal  
0  Barcelona     Basketball       Basketball Men's Basketball   NaN  
1     London           Judo      Judo Men's Extra-Lightweight   NaN  
2  Antwerpen       Football           Football Men's Football   NaN  
3      Paris     Tug-Of-War       Tug-Of-War Men's Tug-Of-War  Gold  
4    Calgary  Speed Skating  Speed Skating Women's 500 metres   NaN

### Four sports to study, basketball, football, ice hockey, and boxing.

In [2]:
# Create a dataframe for the sports of interest
df_sports = df_athlete[df_athlete['Sport'].isin(['Basketball', 'Football', 'Ice Hockey', 'Boxing'])]
df_sports.reset_index(drop=True, inplace=True)
df_sports.head()

Unnamed: 0  ID                                               Name Sex  \
0           0   1  3a4eef48434c66b3f14ab0221f6762d0ef7c6135ab2790...   M   
1           2   3  9c198b205332c2c8e1542e0f9534b9e270780a41d978ec...   M   
2          28   9  a11b41b7e02d7d5ffa770946f8e9cc7f793eefdf62dd90...   M   
3          40  16  ae93a503a6f4349bbb966008ad5ed852094c75c4ef1d08...   M   
4         105  37  6ad689f90c7f37c07ea0f681e85f1bbe8621514ff6f5f3...   F   

    Age  Height  Weight     Team  NOC        Games  Year  Season  \
0  24.0   180.0    80.0    China  CHN  1992 Summer  1992  Summer   
1  24.0     NaN     NaN  Denmark  DEN  1920 Summer  1920  Summer   
2  26.0   186.0    96.0  Finland  FIN  2002 Winter  2002  Winter   
3  28.0   184.0    85.0  Finland  FIN  2014 Winter  2014  Winter   
4  23.0   182.0    64.0   Norway  NOR  1996 Summer  1996  Summer   

             City       Sport                        Event   Medal  
0       Barcelona  Basketball  Basketball Men's Basketball     NaN  
1       Antwerpen    Football      Football Men's Football     NaN  
2  Salt Lake City  Ice Hockey  Ice Hockey Men's Ice Hockey     NaN  
3           Sochi  Ice Hockey  Ice Hockey Men's Ice Hockey  Bronze  
4         Atlanta    Football    Football Women's Football  Bronze

Note: the Unified Team is the name used for the sports team of the former Soviet Union at the 1992 Winter Olympics in Albertville and the 1992 Summer Olympics in Barcelona
- source: https://en.wikipedia.org/wiki/Unified_Team_at_the_Olympics

In [3]:
# get the percentage of missing values Age and Medal columns
print(f"The percentage of missing values in Medal column is {round(df_sports['Medal'].isnull().sum()/len(df_sports['Medal'])*100, 2)}%")
print(f"The percentage of missing values in Age column is {round(df_sports['Age'].isnull().sum()/len(df_sports['Age'])*100, 2)}%")

The percentage of missing values in Medal column is 77.57%
The percentage of missing values in Age column is 3.97%


- As we can see, that the proportion of missing values in the Age column is about 3.97%.
- After thorough examination, the missing values in Medal column are not missing at random, they are missing because the athlete did not win a medal. We will deal with it throughout the project if needed.

In [4]:
# replace the missing values in age column with the mean age
# mean_age = df_sports['Age'].mean()
# df_sports['Age'].fillna(mean_age, inplace=True)

### Boxing analysis and visualization

In [5]:
df_boxing = df_sports[df_sports["Sport"] == "Boxing"]
df_boxing.head()

Unnamed: 0  ID                                               Name Sex  \
6          134  48  0ac6985ff1cd57288555dc653690963912f4c85b3019f0...   M   
7          147  54  762682d159fc9a859fdc85e77ae45aeaf8f593e75c8c7a...   M   
8          151  57  3455ef83abbb13cb9ab0259886130d439eef0026225ae0...   M   
10         162  66  cde2924dfe52473db1370b21ede6848cde04c609b07b0d...   M   
13         179  77  6e7407999a06307c7ad947556c396977650b721e7b8814...   M   

     Age  Height  Weight      Team  NOC        Games  Year  Season  \
6   24.0     NaN    64.0   Morocco  MAR  2012 Summer  2012  Summer   
7    NaN     NaN     NaN      Iran  IRI  1948 Summer  1948  Summer   
8   24.0   176.0    66.0  Bulgaria  BUL  1988 Summer  1988  Summer   
10   NaN   156.0    48.0     Sudan  SUD  1972 Summer  1972  Summer   
13  27.0   166.0    75.0  Cameroon  CMR  1968 Summer  1968  Summer   

           City   Sport                            Event Medal  
6        London  Boxing  Boxing Men's Light-Welterweight   NaN  
7        London  Boxing         Boxing Men's Heavyweight   NaN  
8         Seoul  Boxing  Boxing Men's Light-Welterweight   NaN  
10       Munich  Boxing           Boxing Men's Flyweight   NaN  
13  Mexico City  Boxing        Boxing Men's Middleweight   NaN

In [6]:
df_boxing["Season"].unique() # boxing only has Summer season

array(['Summer'], dtype=object)

Since we are trying to visualize the medal distribution for each land(not athlete), we have to exclude the athletes who did not win a medal.

In [7]:
# remove all columns except Age, Team and Medal
df_boxing = df_boxing[["Age","Sex", "Team", "Year", "Medal"]]

# remove all rows where the medal is NaN
df_boxing = df_boxing[df_boxing["Medal"].notna()]
df_boxing

Age Sex          Team  Year   Medal
69     26.0   M    Uzbekistan  2000    Gold
134    21.0   M   Puerto Rico  1992  Bronze
140    29.0   M       Morocco  1988  Bronze
142    27.0   M       Morocco  1992  Bronze
149    23.0   M          Cuba  2008  Bronze
...     ...  ..           ...   ...     ...
22788  27.0   M         China  2008    Gold
22789  31.0   M         China  2012    Gold
22813  20.0   M         Italy  1948  Silver
22821  22.0   M  East Germany  1988    Gold
22836  21.0   M       Belarus  2004  Silver

[944 rows x 5 columns]

In [8]:
# got this from https://stackoverflow.com/questions/39778686/pandas-reset-index-after-groupby-value-counts

df_boxing_medal = (
    df_boxing.groupby(["Team", "Medal"]) # use groupby to group the data by Team and Medal
    .size() # size() counts the number of rows in each group
    .reset_index(name="Count") # reset_index(name="Count") resets the index and adds a new column called Count to keep track of the number of medals of each type
    .sort_values(by=["Count"], ascending=False) # sort by count of medals
    .reset_index(drop=True) 
)
df_boxing_medal


Team   Medal  Count
0    United States    Gold     50
1    United States  Bronze     39
2             Cuba    Gold     37
3           Poland  Bronze     26
4    Great Britain  Bronze     25
..             ...     ...    ...
167         Guyana  Bronze      1
168          Ghana  Silver      1
169        Georgia  Bronze      1
170        Finland  Silver      1
171         Zambia  Bronze      1

[172 rows x 3 columns]

In [9]:
# plot medal distribution by team
fig = px.bar(
    df_boxing_medal,
    x="Team",
    y="Count",
    log_y=True,
    color="Medal",
    category_orders={"Medal": ["Gold", "Silver", "Bronze"]}, # sort the medals from highest to lowest
    color_discrete_sequence=["gold", "silver", "brown"], # set the colors of the medals
    title="Medal distribution for boxing",
)
fig.show()

NOTE: Did not find a way to show all teams on the x-axis. Hover over the graph to see the data.

In [10]:
# plot the age distribution in teams
# i went with scatter plot because it is easier to read than a histogram/bar chart
fig = px.scatter(
    df_boxing,
    x="Team",
    y="Age",
    color="Team",
    title="Age distribution in boxing teams",
)
fig.show()

In [11]:
# plot the age distribution in teams using box plot
fig = px.box(
    df_boxing,
    x="Team",
    y="Age",
    color="Team",
    title="Age distribution in boxing teams",
)
fig.show()

### Basketball analysis and visualization

In [12]:
df_basketball = df_sports[df_sports["Sport"] == "Basketball"]
df_basketball.head()

Unnamed: 0   ID                                               Name Sex  \
0            0    1  3a4eef48434c66b3f14ab0221f6762d0ef7c6135ab2790...   M   
11         167   69  c695053d84faff079d10d78173869021dda6e6b7a0cfbe...   F   
23         250  124  5e43ba671924351ecca5af521475c999d36ecc78a5c8b4...   M   
25         264  136  482329566fa38a003989f68443e619419467401109e39f...   M   
33         346  192  f3797f608250eea358aaec561a85fce3e8c2248c1a19f3...   M   

     Age  Height  Weight   Team  NOC        Games  Year  Season       City  \
0   24.0   180.0    80.0  China  CHN  1992 Summer  1992  Summer  Barcelona   
11  19.0   185.0    72.0  Spain  ESP  2008 Summer  2008  Summer    Beijing   
23  31.0     NaN     NaN  Egypt  EGY  1952 Summer  1952  Summer   Helsinki   
25  29.0   195.0    85.0  Italy  ITA  2000 Summer  2000  Summer     Sydney   
33  25.0   189.0    85.0  Egypt  EGY  1972 Summer  1972  Summer     Munich   

         Sport                          Event Medal  
0   Basketball    Basketball Men's Basketball   NaN  
11  Basketball  Basketball Women's Basketball   NaN  
23  Basketball    Basketball Men's Basketball   NaN  
25  Basketball    Basketball Men's Basketball   NaN  
33  Basketball    Basketball Men's Basketball   NaN

In [13]:
# remove all nan rows in the Medal column
df_basketball = df_basketball[df_basketball["Medal"].notna()]
df_basketball

Unnamed: 0      ID                                               Name  \
73            609     351  ed1b84400241f3a3cf0028d136bebf5fc81f65fdcd048c...   
122          1029     576  8a9f9c7379a7e8954319e5bbd888b9400c765c663f3feb...   
124          1035     580  92f7e034d79851cf78387c82cabdb2edf994b4d6d68ea2...   
152          1293     727  c1b1ed4d9435ae0f960e3e8dc03d2fa768a2c853bea77c...   
206          1807    1000  7263a55acf8ab546e2e4262d4658ca8daf697e1b4eccf2...   
...           ...     ...                                                ...   
22802      270613  135345  58662432751c11bc9f6ed7aad29b6ffc563a633f9422c3...   
22803      270614  135345  58662432751c11bc9f6ed7aad29b6ffc563a633f9422c3...   
22816      270740  135401  cc6ea64e631e1c9b9a401872e98d8948b405cd0801ce09...   
22817      270741  135401  cc6ea64e631e1c9b9a401872e98d8948b405cd0801ce09...   
22819      270743  135402  daef00596f384dbdd5a0e737eecb87ae3e9e3d6eea1f12...   

      Sex   Age  Height  Weight           Team  NOC        Games  Year  \
73      M  23.0   202.0   104.0  United States  USA  2000 Summer  2000   
122     M  23.0   198.0    93.0          Spain  ESP  2016 Summer  2016   
124     F  28.0   188.0    77.0         Russia  RUS  2008 Summer  2008   
152     M  27.0     NaN     NaN        Uruguay  URU  1952 Summer  1952   
206     M  26.0   202.0   100.0      Lithuania  LTU  2000 Summer  2000   
...    ..   ...     ...     ...            ...  ...          ...   ...   
22802   M  19.0   202.0    88.0   Soviet Union  URS  1956 Summer  1956   
22803   M  23.0   202.0    88.0   Soviet Union  URS  1960 Summer  1960   
22816   M  22.0   218.0   115.0      Lithuania  LTU  1996 Summer  1996   
22817   M  27.0   218.0   115.0      Lithuania  LTU  2000 Summer  2000   
22819   M  20.0   201.0   100.0      Lithuania  LTU  1996 Summer  1996   

       Season            City       Sport                          Event  \
73     Summer          Sydney  Basketball    Basketball Men's Basketball   
122    Summer  Rio de Janeiro  Basketball    Basketball Men's Basketball   
124    Summer         Beijing  Basketball  Basketball Women's Basketball   
152    Summer        Helsinki  Basketball    Basketball Men's Basketball   
206    Summer          Sydney  Basketball    Basketball Men's Basketball   
...       ...             ...         ...                            ...   
22802  Summer       Melbourne  Basketball    Basketball Men's Basketball   
22803  Summer            Roma  Basketball    Basketball Men's Basketball   
22816  Summer         Atlanta  Basketball    Basketball Men's Basketball   
22817  Summer          Sydney  Basketball    Basketball Men's Basketball   
22819  Summer         Atlanta  Basketball    Basketball Men's Basketball   

        Medal  
73       Gold  
122    Bronze  
124    Bronze  
152    Bronze  
206    Bronze  
...       ...  
22802  Silver  
22803  Silver  
22816  Bronze  
22817  Bronze  
22819  Bronze  

[1080 rows x 16 columns]

Since we are doing analysis on each sport separately, we're going to tackle each missing value in every sport accordingly. Because the proportion may vary from sport to sport.

In [14]:
# get the percentage of height and weight missing values
print(f"The percentage of missing values in Height column is {round(df_basketball['Height'].isnull().sum()/len(df_basketball['Height'])*100, 2)}%")
print(f"The percentage of missing values in Weight column is {round(df_basketball['Weight'].isnull().sum()/len(df_basketball['Weight'])*100, 2)}%")

The percentage of missing values in Height column is 5.19%
The percentage of missing values in Weight column is 7.41%


In [15]:
# replace all nan values in height and weight with the mean
# df_basketball["Height"].fillna(df_basketball["Height"].mean(), inplace=True)
# df_basketball["Weight"].fillna(df_basketball["Weight"].mean(), inplace=True)

# # convert height and weight to int
# df_basketball["Height"] = df_basketball["Height"].astype(int)
# df_basketball["Weight"] = df_basketball["Weight"].astype(int)

In [16]:
# plot medal distribution for each team
fig = px.histogram(
    df_basketball,
    x="Team",
    log_y=True,
    barmode="group",
    color="Medal",
    category_orders={"Medal": ["Gold", "Silver", "Bronze"]},
    color_discrete_sequence=["gold", "silver", "brown"],
    title="Medals won by each country in basketball",
)
fig.show()

In [17]:
# age distribution in basketball
# px.box or px.scatter are the best options here, otherwise i'm getting the sum of ages, which is irrelevant
fig = px.box(
    df_basketball,
    x="Team",
    y="Age",
    color="Team",
    title="Age distribution in basketball teams",
)
fig.show()

### Football analysis and visualization

In [18]:
df_football = df_sports[df_sports["Sport"] == "Football"]
df_football.head()

Unnamed: 0  ID                                               Name Sex  \
1            2   3  9c198b205332c2c8e1542e0f9534b9e270780a41d978ec...   M   
4          105  37  6ad689f90c7f37c07ea0f681e85f1bbe8621514ff6f5f3...   F   
5          116  41  aa20025f80eef8028d6cd4928d7314a8c90c4265d56a62...   M   
9          157  61  fdef14e0c91950d56fd8728f20e2e3c4a02ece5485cb85...   M   
12         178  76  249b67de44889b5463368af8135994be238b90cf1d16e1...   M   

     Age  Height  Weight      Team  NOC        Games  Year  Season       City  \
1   24.0     NaN     NaN   Denmark  DEN  1920 Summer  1920  Summer  Antwerpen   
4   23.0   182.0    64.0    Norway  NOR  1996 Summer  1996  Summer    Atlanta   
5   28.0     NaN     NaN    Norway  NOR  1920 Summer  1920  Summer  Antwerpen   
9   26.0   175.0    72.0  Bulgaria  BUL  1960 Summer  1960  Summer       Roma   
12  22.0   185.0    82.0  Cameroon  CMR  2000 Summer  2000  Summer     Sydney   

       Sport                      Event   Medal  
1   Football    Football Men's Football     NaN  
4   Football  Football Women's Football  Bronze  
5   Football    Football Men's Football     NaN  
9   Football    Football Men's Football     NaN  
12  Football    Football Men's Football    Gold

In [19]:
# medal distribution for each team
fig = px.histogram(
    df_football.sort_values(by=["Medal"], ascending=False),
    x="Team",
    log_y=True,
    barmode="group",
    color="Medal",
    category_orders={"Medal": ["Gold", "Silver", "Bronze"]},
    color_discrete_sequence=["gold", "silver", "brown"],
    title="Medals won by each country in football",
)
fig.show()

Here i chose to plot only the teams that won medals, otherwise the plot would be too crowded and hard to read
- In order to plot all teams, just remove the '[df_football["Medal"].notna()]' part

In [20]:
# plot the age distribution in teams
fig = px.box(
    df_football[df_football["Medal"].notna()],
    x="Team",
    y="Age",
    color="Team",
    title="Age distribution in football teams",
)
fig.show()

### Ice Hockey analysis and visualization

In [21]:
df_ice_hockey = df_sports[df_sports["Sport"] == "Ice Hockey"]
df_ice_hockey.head()

Unnamed: 0   ID                                               Name Sex  \
2            28    9  a11b41b7e02d7d5ffa770946f8e9cc7f793eefdf62dd90...   M   
3            40   16  ae93a503a6f4349bbb966008ad5ed852094c75c4ef1d08...   M   
85          672  391  a922ef5f04268dd92156d7fba10b5c03e8a523d968d55b...   M   
86          673  392  c460be799f4065bfcaf54577a8a9ccbbf4de47eff252aa...   M   
111         923  523  301db93827df209fc2574ffc777b8a85a70bda0132e7de...   M   

      Age  Height  Weight           Team  NOC        Games  Year  Season  \
2    26.0   186.0    96.0        Finland  FIN  2002 Winter  2002  Winter   
3    28.0   184.0    85.0        Finland  FIN  2014 Winter  2014  Winter   
85   23.0   185.0   102.0  United States  USA  1924 Winter  1924  Winter   
86   35.0     NaN     NaN         Canada  CAN  1952 Winter  1952  Winter   
111  19.0   183.0    87.0         Norway  NOR  1980 Winter  1980  Winter   

               City       Sport                        Event   Medal  
2    Salt Lake City  Ice Hockey  Ice Hockey Men's Ice Hockey     NaN  
3             Sochi  Ice Hockey  Ice Hockey Men's Ice Hockey  Bronze  
85         Chamonix  Ice Hockey  Ice Hockey Men's Ice Hockey  Silver  
86             Oslo  Ice Hockey  Ice Hockey Men's Ice Hockey    Gold  
111     Lake Placid  Ice Hockey  Ice Hockey Men's Ice Hockey     NaN

In [22]:
# medal distribution for each team
fig = px.histogram(
    df_ice_hockey,
    x="Team",
    log_y=True,
    barmode="group",
    color="Medal",
    category_orders={"Medal": ["Gold", "Silver", "Bronze"]},
    color_discrete_sequence=["gold", "silver", "brown"],
    title="Medals won by each country in ice hockey",
)
fig.show()

In [23]:
# percentage of missing values in height and weight
print(f"The percentage of missing values in Height column is {round(df_ice_hockey['Height'].isnull().sum()/len(df_ice_hockey['Height'])*100, 2)}%")
print(f"The percentage of missing values in Weight column is {round(df_ice_hockey['Weight'].isnull().sum()/len(df_ice_hockey['Weight'])*100, 2)}%")

The percentage of missing values in Height column is 16.41%
The percentage of missing values in Weight column is 16.73%


In [24]:
# replace all nan values in height and weight with the mean
# df_ice_hockey["Height"].fillna(df_ice_hockey["Height"].mean(), inplace=True)
# df_ice_hockey["Weight"].fillna(df_ice_hockey["Weight"].mean(), inplace=True)

In [25]:
# plot the age distribution in teams
fig = px.box(
    df_ice_hockey,
    x="Team",
    y="Age",
    color="Team",
    title="Age distribution in ice hockey teams",
)
fig.show()

## Analysis and visualization of other aspects of these sports

### Top 10 countries with most gold medals

In [26]:
df_top_10_countries = df_sports[df_sports["Medal"] == "Gold"] # get all gold medals
df_top_10_countries = df_top_10_countries["Team"].value_counts().head(10) # return a serie of top 10 countries with most gold medals
df_top_10_countries = pd.DataFrame(df_top_10_countries) # convert to dataframe
df_top_10_countries = df_top_10_countries.reset_index() # reset index
df_top_10_countries.columns = ["Country", "Gold Medals"] # rename columns
df_top_10_countries

Country  Gold Medals
0  United States          453
1   Soviet Union          229
2         Canada          223
3         Sweden           58
4        Hungary           56
5  Great Britain           55
6      Argentina           53
7           Cuba           37
8   Unified Team           34
9        Uruguay           31

In [27]:
fig = px.bar(
    df_top_10_countries,
    x="Country",
    y="Gold Medals",
    color="Country",
    title="Top 10 countries with most gold medals",
)
fig.show()

### Comparison of the number of medals between the USA and the rest of the world

In [28]:
df_usa = df_sports[df_sports["Team"] == "United States"] 
df_usa = df_usa["Year"].value_counts().sort_index()
df_usa = pd.DataFrame(df_usa) 
df_usa = df_usa.reset_index()
df_usa.columns = ["Year", "Medals"]

# create a dataframe with all countries excluding USA
df_world = df_sports[df_sports["Team"] != "United States"]
df_world = df_world["Year"].value_counts().sort_index()
df_world = pd.DataFrame(df_world)
df_world = df_world.reset_index() 
df_world.columns = ["Year", "Medals"]

In [29]:
# using plotly.graph_objects turns out to be easier to use when comparing from two different dataframes
# https://stackoverflow.com/questions/60372991/plotly-how-to-plot-two-lines-from-two-dataframe-columns-and-assign-hover-info-f

fig = go.Figure()
fig.add_trace(go.Scatter(x=df_usa["Year"], y=df_usa["Medals"], name="USA"))
fig.add_trace(go.Scatter(x=df_world["Year"], y=df_world["Medals"], name="World"))
fig.update_layout(title="Medals won by the USA vs the world over time")
fig.show()

### Number of minors in each sport

In [30]:
# plot the number of minors in each sport
df_minor = df_sports[df_sports["Age"] < 18]
df_minor = df_minor["Sport"].value_counts()
df_minor = pd.DataFrame(df_minor)
df_minor = df_minor.reset_index()
df_minor.columns = ["Sport", "Number of minors"]

fig = px.bar(
    df_minor,
    x="Sport",
    y="Number of minors",
    color="Sport",
    title="Number of minors in each sport",
)
fig.show()

### Weight and height average in each sport

In [31]:
df_weight = df_sports.groupby("Sport")["Weight"].mean()
df_weight = pd.DataFrame(df_weight)
df_weight = df_weight.reset_index()
df_weight.columns = ["Sport", "Average weight"]

df_height = df_sports.groupby("Sport")["Height"].mean()
df_height = pd.DataFrame(df_height)
df_height = df_height.reset_index()
df_height.columns = ["Sport", "Average height"]

df_height_weight = pd.merge(df_weight, df_height, on="Sport")
df_height_weight

fig = px.bar(
    df_height_weight,
    x="Sport",
    y=["Average weight", "Average height"],
    barmode="group",
    title="Average weight and height in each sport",
)
fig.show()